### ISSUE TITLE
SSL_free doesn't free write buffer because of SSL_clear

### ISSUE THREAD

**Issue by kaichieh** [openssl/openssl#2026]
Version:1.1.0
The sample code can reproduce the problem.
I know I don't need to call SSL_clear if there is no new connection. 
It look like SSL_clear clear the flag so SSL_free doesn't free the write buffer.
Or I misunderstand the usage of SSL_clear?
The following is the sample code.

```c
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <netdb.h>
#include "openssl/ssl.h"


int main(int argc, char **argv){
    if(argc <=2)
        printf("usage:%s [host] [port]\n", argv[0]);
    int error_code,status;
    struct addrinfo hints;
    struct addrinfo *serverInfo;
    memset(&hints, 0 , sizeof(hints));
    hints.ai_family = AF_UNSPEC;
    hints.ai_socktype = SOCK_STREAM;
    if(status = getaddrinfo(argv[1],argv[2],&hints, &serverInfo) !=0){
        printf("fail to get address of %s", argv[1]);
        exit(-1);
    }
    int sockfd = socket(serverInfo->ai_family, serverInfo->ai_socktype, serverInfo->ai_protocol);
    if(sockfd < 0){
        printf("fail to create socket");
        exit(-1);
    }

    if(status = connect(sockfd, serverInfo->ai_addr, serverInfo->ai_addrlen) != 0){
        printf("fail to connect to %s", argv[1]);
        exit(-1);
    }

    const SSL_METHOD * method = TLS_client_method();
    if(NULL == method){
        printf("fail to create method");
        exit(-1);
    }

    SSL_CTX * sslCtx = SSL_CTX_new(method);
    if(NULL == sslCtx){
        printf("fail to create sslCtx");
        exit(-1);
    }

    SSL* ssl = SSL_new(sslCtx);
    if(NULL == ssl){
        printf("fail to create ssl");
        exit(-1);
    }
    if(0==SSL_set_fd(ssl, sockfd)){
        printf("fail to set fd to openssl");
        exit(-1);
    }
    if((error_code = SSL_connect(ssl))!=1){
        printf("fail to handshark%d %d\n",error_code, SSL_get_error(ssl,error_code));
        exit(-1);
    }

    if((error_code = SSL_shutdown(ssl))<0){
        printf("fail to handshark %d %d\n",error_code, SSL_get_error(ssl,error_code));
        exit(-1);
    }

    if(0==SSL_clear(ssl)){
        printf("fail to clear");
        exit(-1);
    }
    SSL_free(ssl);
    SSL_CTX_free(sslCtx);
    freeaddrinfo(serverInfo);
    close(sockfd);
    return 0;
}
```
valgrind message
==32005== 
==32005== 16,560 bytes in 1 blocks are definitely lost in loss record 1 of 1
==32005==    at 0x4A069EE: malloc (vg_replace_malloc.c:270)
==32005==    by 0x40D625: ssl3_setup_write_buffer (ssl3_buffer.c:109)
==32005==    by 0x40D6AE: ssl3_setup_buffers (ssl3_buffer.c:129)
==32005==    by 0x41CF18: state_machine (statem.c:314)
==32005==    by 0x418B40: SSL_do_handshake (ssl_lib.c:3025)
==32005==    by 0x40938F: main (in /home/Jay/ssl_test/a.out)
==32005== 
==32005== LEAK SUMMARY:
==32005==    definitely lost: 16,560 bytes in 1 blocks
==32005==    indirectly lost: 0 bytes in 0 blocks
==32005==      possibly lost: 0 bytes in 0 blocks
==32005==    still reachable: 0 bytes in 0 blocks
==32005==         suppressed: 0 bytes in 0 blocks

**Comment by richsalz** [comment]
This looks like a bug in state machine stuff; ping @mattcaswell

**Comment by mattcaswell** [comment]
See fix in #2032

**Comment by mattcaswell** [comment]
Fixed in master and 1.1.0
